ChiModules/LinearBoltzmannSolver/lbs_compute_balance.cc

```diff
@@ -0,0 +1,167 @@
+#include "lbs_linear_boltzmann_solver.h"
+
+#include "FiniteElement/PiecewiseLinear/pwl.h"
+
+#include "chi_log.h"
+extern ChiLog& chi_log;
+
+#include <iomanip>
+
+//###################################################################
+/**Zeroes all the outflow data-structures required to compute
+ * balance.*/
+void LinearBoltzmann::Solver::ZeroOutflowBalanceVars(LBSGroupset& groupset)
+{
+  for (auto& cell_transport_view : cell_transport_views)
+    for (auto& group : groupset.groups)
+      cell_transport_view.ZeroOutflow(group.id);
+}
+
+//###################################################################
+/**Compute balance.*/
+void LinearBoltzmann::Solver::ComputeBalance()
+{
+  MPI_Barrier(MPI_COMM_WORLD);
+  chi_log.Log() << "\n********** Computing balance\n";
+
+  auto pwld =
+    std::dynamic_pointer_cast<SpatialDiscretization_PWLD>(discretization);
+  if (not pwld) throw std::logic_error("Trouble getting PWLD-SDM in " +
+                                      std::string(__FUNCTION__));
+  SpatialDiscretization_PWLD& grid_fe_view = *pwld;
+
+  //======================================== Get material source
+  // This is done using the SetSource routine
+  // because it allows a lot of flexibility.
+  auto mat_src = phi_old_local;
+  mat_src.assign(mat_src.size(),0.0);
+  for (auto& groupset : group_sets)
+  {
+    SetSource(groupset, APPLY_MATERIAL_SOURCE | APPLY_FISSION_SOURCE);
+    ScopedCopySTLvectors(groupset, q_moments_local, mat_src);
+  }
+
+  //======================================== Initialize diffusion params
+  //                                         for xs
+  // This populates sigma_a
+  for (auto& xs : material_xs)
+    if (not xs->diffusion_initialized)
+      xs->ComputeDiffusionParameters();
+
+  //======================================== Compute absorption, material-source
+  //                                         and in-flow
+  size_t num_groups=groups.size();
+  double local_out_flow   = 0.0;
+  double local_in_flow    = 0.0;
+  double local_absorption = 0.0;
+  double local_production = 0.0;
+  for (const auto& cell : grid->local_cells)
+  {
+    const auto&  transport_view   = cell_transport_views[cell.local_id];
+    const auto&  fe_intgrl_values = grid_fe_view.GetUnitIntegrals(cell);
+    const size_t num_nodes        = transport_view.NumNodes();
+    const auto&  IntV_shapeI      = fe_intgrl_values.GetIntV_shapeI();
+    const auto&  IntS_shapeI      = fe_intgrl_values.GetIntS_shapeI();
+
+    //====================================== Inflow
+    // This is essentially an integration over
+    // all faces, all angles, and all groups.
+    // Only the cosines that are negative are
+    // added to the integral.
+    for (int f=0; f<cell.faces.size(); ++f)
```

Comment:
This will probably generate a compiler warning if size() is returning a size_t.  I'd make f size_t as well.

Reply:
Done

Reply:
Is this an across the board thing always? Now I see it everywhere and neither the compiler or clang-tidy complains about it? 

Does it also mean that whenever we have an STL container that `for (int` is taboo?

Reply:
Merely an observation: I have noticed various concerns arising from inconsistency of types between loop counters and loop limits, but why are there no concerns regarding inconsistency of types arising from the use of the loop counter within the body of the loop (e.g. the signature of the function to which it might be passed)? I mention this as we have many instances where ensuring consistency between any two of the three results in an inconsistency with the third. To me, the ideal is to be entirely consistent, but at the same time, I think that implicit type conversion is handling some of the burden.

Reply:
I suspect Dominic is correct about implicit type conversion.  My $0.02 on our signed/unsigned/int/size_t issues.  Unless the quantity represented is unlikely to fit into 32 bits, use an int.  This is simple, consistent, and natural from a coding perspective.  If the quantity will exceed the value that can be represented by 32 bits, using an int64_t or uint64_t (likely only necessary for global ids).  IMHO, size_t was an unfortunately choice by the maintainers, but I see why they did it... they really didn't have much choice.  Use size_t where needed (looping over container sizes say) and cast to int as necessary.